File: voca_crm/lib/domain/entity/error_log.dart
/// 오류 로그 엔티티
class ErrorLog {
  final String id;
  final String? userId;
  final String? username;
  final String? businessPlaceId;
  final String? screenName;
  final String? action;
  final String? requestUrl;
  final String? requestMethod;
  final String? requestBody;
  final int? httpStatusCode;
  final String? errorCode;
  final String errorMessage;
  final String? stackTrace;
  final ErrorSeverity severity;
  final String? deviceInfo;
  final String? appVersion;
  final String? osVersion;
  final String? platform;
  final bool resolved;
  final String? resolvedBy;
  final DateTime? resolvedAt;
  final String? resolutionNote;
  final DateTime createdAt;

  ErrorLog({
    required this.id,
    this.userId,
    this.username,
    this.businessPlaceId,
    this.screenName,
    this.action,
    this.requestUrl,
    this.requestMethod,
    this.requestBody,
    this.httpStatusCode,
    this.errorCode,
    required this.errorMessage,
    this.stackTrace,
    this.severity = ErrorSeverity.error,
    this.deviceInfo,
    this.appVersion,
    this.osVersion,
    this.platform,
    this.resolved = false,
    this.resolvedBy,
    this.resolvedAt,
    this.resolutionNote,
    required this.createdAt,
  });

  factory ErrorLog.fromJson(Map<String, dynamic> json) {
    return ErrorLog(
      id: json['id'] as String,
      userId: json['userId'] as String?,
      username: json['username'] as String?,
      businessPlaceId: json['businessPlaceId'] as String?,
      screenName: json['screenName'] as String?,
      action: json['action'] as String?,
      requestUrl: json['requestUrl'] as String?,
      requestMethod: json['requestMethod'] as String?,
      requestBody: json['requestBody'] as String?,
      httpStatusCode: json['httpStatusCode'] as int?,
      errorCode: json['errorCode'] as String?,
      errorMessage: json['errorMessage'] as String,
      stackTrace: json['stackTrace'] as String?,
      severity: ErrorSeverity.fromString(json['severity'] as String?),
      deviceInfo: json['deviceInfo'] as String?,
      appVersion: json['appVersion'] as String?,
      osVersion: json['osVersion'] as String?,
      platform: json['platform'] as String?,
      resolved: json['resolved'] as bool? ?? false,
      resolvedBy: json['resolvedBy'] as String?,
      resolvedAt: json['resolvedAt'] != null
          ? DateTime.parse(json['resolvedAt'] as String)
          : null,
      resolutionNote: json['resolutionNote'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'username': username,
      'businessPlaceId': businessPlaceId,
      'screenName': screenName,
      'action': action,
      'requestUrl': requestUrl,
      'requestMethod': requestMethod,
      'requestBody': requestBody,
      'httpStatusCode': httpStatusCode,
      'errorCode': errorCode,
      'errorMessage': errorMessage,
      'stackTrace': stackTrace,
      'severity': severity.name.toUpperCase(),
      'deviceInfo': deviceInfo,
      'appVersion': appVersion,
      'osVersion': osVersion,
      'platform': platform,
      'resolved': resolved,
      'resolvedBy': resolvedBy,
      'resolvedAt': resolvedAt?.toIso8601String(),
      'resolutionNote': resolutionNote,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

/// 오류 심각도
enum ErrorSeverity {
  info('INFO', '정보'),
  warning('WARNING', '경고'),
  error('ERROR', '오류'),
  critical('CRITICAL', '치명적');

  final String value;
  final String displayName;

  const ErrorSeverity(this.value, this.displayName);

  static ErrorSeverity fromString(String? value) {
    if (value == null) return ErrorSeverity.error;
    return ErrorSeverity.values.firstWhere(
      (e) => e.value == value.toUpperCase(),
      orElse: () => ErrorSeverity.error,
    );
  }
}

/// 오류 로그 페이지네이션 응답
class ErrorLogPage {
  final List<ErrorLog> content;
  final int totalElements;
  final int totalPages;
  final int number; // current page
  final int size;

  ErrorLogPage({
    required this.content,
    required this.totalElements,
    required this.totalPages,
    required this.number,
    required this.size,
  });

  factory ErrorLogPage.fromJson(Map<String, dynamic> json) {
    return ErrorLogPage(
      content: (json['content'] as List<dynamic>)
          .map((e) => ErrorLog.fromJson(e as Map<String, dynamic>))
          .toList(),
      totalElements: json['totalElements'] as int,
      totalPages: json['totalPages'] as int,
      number: json['number'] as int,
      size: json['size'] as int,
    );
  }
}

/// 오류 통계 요약
class ErrorSummary {
  final int totalErrors;
  final int unresolvedErrors;
  final Map<String, int> bySeverity;
  final List<ScreenErrorStat> byScreen;

  ErrorSummary({
    required this.totalErrors,
    required this.unresolvedErrors,
    required this.bySeverity,
    required this.byScreen,
  });

  factory ErrorSummary.fromJson(Map<String, dynamic> json) {
    final bySeverityJson = json['bySeverity'] as Map<String, dynamic>? ?? {};
    final byScreenJson = json['byScreen'] as List<dynamic>? ?? [];

    return ErrorSummary(
      totalErrors: json['totalErrors'] as int? ?? 0,
      unresolvedErrors: json['unresolvedErrors'] as int? ?? 0,
      bySeverity: bySeverityJson.map((k, v) => MapEntry(k, (v as num).toInt())),
      byScreen: byScreenJson
          .map((e) => ScreenErrorStat.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

/// 화면별 오류 통계
class ScreenErrorStat {
  final String screenName;
  final int errorCount;

  ScreenErrorStat({
    required this.screenName,
    required this.errorCount,
  });

  factory ScreenErrorStat.fromJson(Map<String, dynamic> json) {
    return ScreenErrorStat(
      screenName: json['screenName'] as String? ?? 'Unknown',
      errorCount: (json['errorCount'] as num?)?.toInt() ?? 0,
    );
  }
}

File: voca_crm_api/src/main/java/com/vocacrm/api/model/ErrorLog.java
package com.vocacrm.api.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * 오류 로그 (Error Log) 엔티티
 *
 * 클라이언트(Flutter 앱)에서 발생한 오류를 기록합니다.
 * - 누가 (userId)
 * - 언제 (createdAt)
 * - 어느 화면에서 (screenName)
 * - 어떤 행동으로 (action)
 * - 어떤 요청으로 (requestUrl, requestMethod, requestBody)
 * - 어떤 오류가 (errorMessage, stackTrace)
 * - 어떤 기기에서 (deviceInfo, appVersion, osVersion)
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "error_logs", indexes = {
        @Index(name = "idx_error_user_id", columnList = "user_id"),
        @Index(name = "idx_error_created_at", columnList = "created_at"),
        @Index(name = "idx_error_business_place", columnList = "business_place_id"),
        @Index(name = "idx_error_severity", columnList = "severity"),
        @Index(name = "idx_error_screen", columnList = "screen_name"),
        @Index(name = "idx_error_resolved", columnList = "resolved")
})
public class ErrorLog {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    /**
     * 오류가 발생한 사용자 ID (비로그인 상태면 null)
     */
    @Column(name = "user_id")
    private UUID userId;

    /**
     * 사용자 이름 (조회 편의를 위해 비정규화)
     */
    @Column(name = "username", length = 100)
    private String username;

    /**
     * 사업장 ID
     */
    @Column(name = "business_place_id", length = 7)
    private String businessPlaceId;

    /**
     * 오류가 발생한 화면 이름
     */
    @Column(name = "screen_name", length = 100)
    private String screenName;

    /**
     * 사용자가 수행하던 행동
     */
    @Column(name = "action", length = 100)
    private String action;

    /**
     * API 요청 URL
     */
    @Column(name = "request_url", length = 500)
    private String requestUrl;

    /**
     * HTTP 메서드 (GET, POST, PUT, DELETE 등)
     */
    @Column(name = "request_method", length = 10)
    private String requestMethod;

    /**
     * 요청 본문 (민감 정보 제외)
     */
    @Column(name = "request_body", columnDefinition = "TEXT")
    private String requestBody;

    /**
     * HTTP 응답 상태 코드
     */
    @Column(name = "http_status_code")
    private Integer httpStatusCode;

    /**
     * 오류 코드 (앱에서 정의한 코드)
     */
    @Column(name = "error_code", length = 50)
    private String errorCode;

    /**
     * 오류 메시지
     */
    @Column(name = "error_message", columnDefinition = "TEXT")
    private String errorMessage;

    /**
     * 스택 트레이스
     */
    @Column(name = "stack_trace", columnDefinition = "TEXT")
    private String stackTrace;

    /**
     * 심각도
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "severity", length = 20, nullable = false)
    private ErrorSeverity severity = ErrorSeverity.ERROR;

    /**
     * 디바이스 정보 (모델명 등)
     */
    @Column(name = "device_info", length = 200)
    private String deviceInfo;

    /**
     * 앱 버전
     */
    @Column(name = "app_version", length = 20)
    private String appVersion;

    /**
     * OS 버전 (예: iOS 17.0, Android 14)
     */
    @Column(name = "os_version", length = 50)
    private String osVersion;

    /**
     * 플랫폼 (iOS, Android)
     */
    @Column(name = "platform", length = 20)
    private String platform;

    /**
     * 해결 여부
     */
    @Column(name = "resolved", nullable = false)
    private Boolean resolved = false;

    /**
     * 해결한 관리자 ID
     */
    @Column(name = "resolved_by")
    private UUID resolvedBy;

    /**
     * 해결 시간
     */
    @Column(name = "resolved_at")
    private LocalDateTime resolvedAt;

    /**
     * 해결 메모
     */
    @Column(name = "resolution_note", length = 500)
    private String resolutionNote;

    /**
     * 오류 발생 시간
     */
    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    /**
     * 오류 심각도
     */
    public enum ErrorSeverity {
        INFO,       // 정보 (참고용)
        WARNING,    // 경고 (주의 필요)
        ERROR,      // 오류 (처리 필요)
        CRITICAL    // 치명적 (즉시 처리 필요)
    }
}
